src/expressions/whereExpression.ts

```diff
@@ -0,0 +1,50 @@
+import ISequence from './dataTypes/ISequence';
+import sequenceFactory from './dataTypes/sequenceFactory';
+import Value from './dataTypes/Value';
+import DynamicContext from './DynamicContext';
+import ExecutionParameters from './ExecutionParameters';
+import Expression, { RESULT_ORDERINGS } from './Expression';
+import PossiblyUpdatingExpression from './PossiblyUpdatingExpression';
+import Specificity from './Specificity';
+import { IAsyncIterator, IterationHint, notReady } from './util/iterators';
+
+class WhereExpression extends PossiblyUpdatingExpression {
+	constructor(testExpression: Expression, returnExpression: Expression) {
+		const specificity = new Specificity({});
+		super(specificity, [testExpression, returnExpression], {
+			canBeStaticallyEvaluated: false,
+			peer: false,
+			resultOrder: RESULT_ORDERINGS.UNSORTED,
+			subtree: false
+		});
+	}
+
+	public performFunctionalEvaluation(
+		dynamicContext: DynamicContext,
+		_executionParameters: ExecutionParameters,
+		sequenceCallbacks: ((dynamicContext: DynamicContext) => ISequence)[]
+	) {
+		let resultIterator: IAsyncIterator<Value> | null = null;
+		const whereExpressioResultSequence = sequenceCallbacks[0](dynamicContext);
+
+		return sequenceFactory.create({
+			next: (hint: IterationHint) => {
+				if (!resultIterator) {
+					const whereExpressionResult = whereExpressioResultSequence.tryGetEffectiveBooleanValue();
+
+					if (!whereExpressionResult.ready) {
+						return notReady(whereExpressionResult.promise);
```

Comment:
The 0.002% coverage decrease is because we don't enter here in the test

Reply:
Removing this "if" doesn't break anything, so we can just delete it for good, if everybody agrees.

Reply:
it will if you put async expressions in the where, it will be hit:

```
for $i in (1,2,3)
where fontoxpath:sleep($i) > 1
return $i
```

We are not always testing this properly though, but in this case we can.

Reply:
Ok, I'm going to add the test, thanks